vshard/replicaset.lua

```diff
@@ -75,6 +75,144 @@ local fiber_cond_wait = util.fiber_cond_wait
+    local vconn = conn.vconnect
+    if not vconn then
+        -- Nothing to do. Check is not needed.
+        return
+    end
+
+    local opts = {is_async = true}
+    vconn.future = conn:call('vshard.storage._call', {'info'}, opts)
+    vconn.future_cond:broadcast()
+end
+
+--
+-- Check, that future is ready, and its result is expected.
+-- The function doesn't yield.
+-- @retval true; The correct response is received.
+-- @retval nil, ...; Response is not received or validation error happened.
+--
+local function conn_vconnect_check(conn)
+    local vconn = conn.vconnect
+    -- conn.vconnect may be nil, if connection was created on old version
+    -- and the storage was reloaded to a new one. It's also nil, when
+    -- all checks were already done.
+    if not vconn then
+        return true
+    end
+    -- Nothing to do, but wait in such case.
+    if not vconn.future or not vconn.future:is_ready() then
+        return nil, lerror.vshard(lerror.code.VHANDSHAKE_NOT_COMPLETE,
+                                  conn.replica.id)
+    end
+    -- Critical errors. Connection should be closed after these ones.
+    local result, err = vconn.future:result()
+    if not result then
+        -- Failed to get response. E.g. access error.
+        return nil, lerror.make(err)
+    end
+    if vconn.is_named and result[1].name ~= conn.replica.name then
+        return nil, lerror.vshard(lerror.code.INSTANCE_NAME_MISMATCH,
+                                  conn.replica.name, result[1].name)
+    end
+    -- Don't validate until reconnect happens.
+    conn.vconnect = nil
+    return true
+end
+
+local function conn_vconnect_check_or_close(conn)
+    local ok, err = conn_vconnect_check(conn)
+    -- Close the connection, if error happened, but it is not
+    -- VSHANDSHAKE_NOT_COMPLETE.
+    if not ok and err and not (err.type == 'ShardingError' and
+       err.code == lerror.code.VHANDSHAKE_NOT_COMPLETE) then
+        conn:close()
+    end
+    return ok, err
+end
+
+--
+-- Wait until the future object is ready. Returns remaining timeout.
+-- @retval timeout; Future boject is ready.
+-- @retval nil, err; Timeout passed.
+--
+local function conn_vconnect_wait(conn, timeout)
+    local vconn = conn.vconnect
+    -- Fast path. In most cases no validation should be done at all.
+    -- conn.vconnect may be nil, if connection was created on old version
+    -- and the storage was reloaded to a new one. It's also nil, when
+    -- all checks were already done.
+    if not vconn or (vconn.future and vconn.future:is_ready()) then
+        return timeout
+    end
+    local deadline = fiber_clock() + timeout
+    -- Wait for connection to be established.
+    if not vconn.future and
+       not fiber_cond_wait(vconn.future_cond, timeout) then
+        return nil, lerror.timeout()
+    end
+    timeout = deadline - fiber_clock()
+    -- Wait for async call to return.
+    local res, err = util.future_wait(vconn.future, timeout)
```

Comment:
There is `future_wait` function in replicaset.lua as a variable. You don't need `util.` prefix.

Reply:
Fixed